File: sources/CPP/include/Ers/SubModel/Event/RelationEvents.cpp
#include "RelationEvents.h"

#include "Ers/SubModel/SubModel.h"

#include "Ers/Api.h"

void* Ers::RelationEvents::Data()
{
    return this;
}

const void* const Ers::RelationEvents::Data() const
{
    return this;
}

Ers::EventSignal<Ers::Entity> Ers::RelationEvents::OnAwake()
{

    return Ers::EventSignal<Entity>(
        ersAPIFunctionPointers.ERS_SubModel_RelationEvents_OnAwake(), ersAPIFunctionPointers.ERS_SubModel_RelationEvents_OnAwake_Connect);
}

Ers::EventSignal<Ers::Entity> Ers::RelationEvents::OnStart()
{
    return Ers::EventSignal<Entity>(
        ersAPIFunctionPointers.ERS_SubModel_RelationEvents_OnStart(), ersAPIFunctionPointers.ERS_SubModel_RelationEvents_OnStart_Connect);
}

Ers::EventSignal<Ers::Entity> Ers::RelationEvents::OnCreate()
{
    return Ers::EventSignal<Entity>(
        ersAPIFunctionPointers.ERS_SubModel_RelationEvents_OnCreate(), ersAPIFunctionPointers.ERS_SubModel_RelationEvents_OnCreate_Connect);
}

Ers::EventSignal<Ers::Entity> Ers::RelationEvents::OnDestroy()
{
    return Ers::EventSignal<Entity>(
        ersAPIFunctionPointers.ERS_SubModel_RelationEvents_OnDestroy(),
        ersAPIFunctionPointers.ERS_SubModel_RelationEvents_OnDestroy_Connect);
}

Ers::EventSignal<Ers::Entity, Ers::Entity> Ers::RelationEvents::OnEntering()
{
    return Ers::EventSignal<Entity, Entity>(
        ersAPIFunctionPointers.ERS_SubModel_RelationEvents_OnEntering(),
        ersAPIFunctionPointers.ERS_SubModel_RelationEvents_OnEntering_Connect);
}

Ers::EventSignal<Ers::Entity, Ers::Entity> Ers::RelationEvents::OnEntered()
{
    return Ers::EventSignal<Entity, Entity>(
        ersAPIFunctionPointers.ERS_SubModel_RelationEvents_OnEntered(),
        ersAPIFunctionPointers.ERS_SubModel_RelationEvents_OnEntered_Connect);
}

Ers::EventSignal<Ers::Entity, Ers::Entity> Ers::RelationEvents::OnExiting()
{
    return Ers::EventSignal<Entity, Entity>(
        ersAPIFunctionPointers.ERS_SubModel_RelationEvents_OnExiting(),
        ersAPIFunctionPointers.ERS_SubModel_RelationEvents_OnExiting_Connect);
}

Ers::EventSignal<Ers::Entity, Ers::Entity> Ers::RelationEvents::OnExited()
{
    return Ers::EventSignal<Entity, Entity>(
        ersAPIFunctionPointers.ERS_SubModel_RelationEvents_OnExited(), ersAPIFunctionPointers.ERS_SubModel_RelationEvents_OnExited_Connect);
}

File: sources/CSharp/src/Ers/SubModel/Event/RelationEvents.cs
using Ers.Engine;
using System.Runtime.InteropServices;

#pragma warning disable CS1591 // Missing XML comment for publicly visible type or member
namespace Ers
{
    [Obsolete("Replaced by overridable methods (OnEntering, OnEntered, OnExiting, OnExited) on ScriptBehavior")]
    [StructLayout(LayoutKind.Sequential)]
    public ref struct RelationEvents
    {
        // Default events
        public EventSignal1 OnAwake =>
            new EventSignal1(ErsEngine.ERS_SubModel_RelationEvents_OnAwake(), ErsEngine.ERS_SubModel_RelationEvents_OnAwake_Connect);
        public EventSignal1 OnStart =>
            new EventSignal1(ErsEngine.ERS_SubModel_RelationEvents_OnStart(), ErsEngine.ERS_SubModel_RelationEvents_OnStart_Connect);

        public EventSignal1 OnCreate =>
            new EventSignal1(ErsEngine.ERS_SubModel_RelationEvents_OnCreate(), ErsEngine.ERS_SubModel_RelationEvents_OnCreate_Connect);
        public EventSignal1 OnDestroy =>
            new EventSignal1(ErsEngine.ERS_SubModel_RelationEvents_OnDestroy(), ErsEngine.ERS_SubModel_RelationEvents_OnDestroy_Connect);

        public EventSignal1 OnEntering =>
            new EventSignal1(ErsEngine.ERS_SubModel_RelationEvents_OnEntering(), ErsEngine.ERS_SubModel_RelationEvents_OnEntering_Connect);
        public EventSignal2 OnEntered =>
            new EventSignal2(ErsEngine.ERS_SubModel_RelationEvents_OnEntered(), ErsEngine.ERS_SubModel_RelationEvents_OnEntered_Connect);

        public EventSignal1 OnExiting =>
            new EventSignal1(ErsEngine.ERS_SubModel_RelationEvents_OnExiting(), ErsEngine.ERS_SubModel_RelationEvents_OnExiting_Connect);
        public EventSignal2 OnExited =>
            new EventSignal2(ErsEngine.ERS_SubModel_RelationEvents_OnExited(), ErsEngine.ERS_SubModel_RelationEvents_OnExited_Connect);

        private readonly IntPtr SubmodelPtr;

        internal RelationEvents(IntPtr subModel) { SubmodelPtr = subModel; }
    }
}
#pragma warning restore CS1591 // Missing XML comment for publicly visible type or member
